### PR TITLE
Compiler: lower level switch

### LINKED ISSUE #1495: Compiler: Lower level Code.Switch
`Code.Switch` currently deal with both immediate and block/tagged values.
I feel we could simplify things a bit by adding an explicit condition on `is_int` and have `Switch` just dispatch on a variable immediate value

In other words, I means turning 
```
B1: Switch(x,A1,A2)
```

Into
```
B1: Cond(isInt x, B2, B3)
B2: Switch(x, A1)
B3: let x' = tag x; Switch (x', A2)
```

### PR DESCRIPTION
**Pull request by hhugo**
fix #1495

### CODE REVIEW

File: compiler/lib/global_flow.ml [hunk 1]
```diff
@@ -243,14 +243,9 @@ let program_deps st { blocks; _ } =
...
           cont_deps blocks st cont1;
           cont_deps blocks st ~ignore:x cont2
-      | Switch (x, a1, a2) ->
+      | Switch (x, a1) ->
           Array.iter a1 ~f:(fun cont -> cont_deps blocks st cont);
-          Array.iter a2 ~f:(fun cont -> cont_deps blocks st cont);
           let h = Hashtbl.create 16 in
-          Array.iteri
-            ~f:(fun i (pc, _) ->
-              Hashtbl.replace h pc (i :: (try Hashtbl.find h pc with Not_found -> [])))
```
Comment (hhugo) [inline compiler/lib/global_flow.ml:252]: @vouillon, how important is the optimization about known_cases ? do we need to restore it ?

File: compiler/lib/eval.ml [hunk 2]
```diff
@@ -341,14 +341,13 @@ let eval_branch info (l, loc) =
...
           | Unknown -> b)
-    | Switch (x, const, tags) as b -> (
+    | Switch (x, const) as b -> (
         (* [the_case_of info (Pv x)] might be meaningless when we're inside a dead code.
            The proper fix would be to remove the deadcode entirely.
            Meanwhile, add guards to prevent Invalid_argument("index out of bounds")
            see https://github.com/ocsigen/js_of_ocaml/issues/485 *)
         match the_case_of info (Pv x) with
         | CConst j when j >= 0 && j < Array.length const -> Branch const.(j)
-        | CTag j when j >= 0 && j < Array.length tags -> Branch tags.(j)
```
Comment (hhugo) [inline compiler/lib/eval.ml:351]: This knowledge could be used when evaluating "%direct_obj_tag"

File: compiler/tests-full/stdlib.cma.expected.js [hunk 3]
```diff
@@ -15075,85 +15078,93 @@
...
+            switch$1 = 1; break;
            case 11:
-            switch$0 = 1; break;
+            switch$0 = 1; switch$1 = 1; break;
            case 12:
-            switch$0 = 2; break;
+            switch$0 = 2; switch$1 = 1; break;
            case 13:
-            switch$0 = 3; break;
+            switch$0 = 3; switch$1 = 1; break;
```
Comment (hhugo) [inline compiler/tests-full/stdlib.cma.expected.js:15153]: The diff in this file may look like a regression (in term of code size) but it looks much better once rebased on top of #1496

File: compiler/lib/eval.ml (hunk 2 above)
Comment (hhugo) [reply compiler/lib/eval.ml:351]: done

File: compiler/lib/global_flow.ml (hunk 1 above)
Comment (hhugo) [reply compiler/lib/global_flow.ml:254]: I've tried to restore the optimization but it appears that our test-suite doesn't capture/test this.
Would you be able to contribute an expect test for this ?

Comment (vouillon) [reply compiler/lib/global_flow.ml:254]: I'm not sure it is really useful in practice. It makes the analysis more precise and does not cost much, but it's not clear to me whether it make much of a difference.

Comment (vouillon) [reply compiler/lib/global_flow.ml:254]: Your change looks fine. I need to think of a test.

**Pull request closed without merging**